app/preparation.py

```diff
@@ -0,0 +1,25 @@
+def preparation(func):
+    def inner(knights: dict):
+        for knight, attributes in knights.items():
+
+            # apply weapon
+            attributes["power"] += attributes["weapon"]["power"]
+
+            # apply armour
+            attributes["protection"] = 0
+            if attributes["armour"] != []:
+                for a in attributes["armour"]:
+                    attributes["protection"] += a["protection"]
+
+            # apply potion if exist
+            if attributes["potion"] is not None:
+                if "power" in attributes["potion"]["effect"]:
+                    attributes["power"] += attributes["potion"]["effect"]["power"]
+
+                if "protection" in attributes["potion"]["effect"]:
+                    attributes["protection"] += attributes["potion"]["effect"]["protection"]
+
+                if "hp" in attributes["potion"]["effect"]:
+                    attributes["hp"] += attributes["potion"]["effect"]["hp"]
```

Comment:
Use loop here. As you may see the operations are repeated. You can use setattr and getattr functions to avoid code duplicates